packages/flutter_tools/lib/src/compile.dart

```diff
@@ -0,0 +1,138 @@
+import 'package:flutter_tools/src/base/file_system.dart';
+import 'package:flutter_tools/src/base/io.dart';
+import 'package:flutter_tools/src/globals.dart';
+import 'package:usage/uuid/uuid.dart';
+
+String _dartExecutable() {
+  final String engineDartSdkPath = artifacts.getArtifactPath(
+    Artifact.engineDartSdkPath
+  );
+  return fs.path.join(engineDartSdkPath, 'bin', 'dart');
+}
+
+Future<String> compile({String sdkRoot, String mainPath}) async {
+  final String frontendServer = artifacts.getArtifactPath(
+    Artifact.frontendServerSnapshotForEngineDartSdk
+  );
+
+  if (!sdkRoot.endsWith('/'))
+    sdkRoot = "$sdkRoot/";
+  final Process server = await Process.start(_dartExecutable(),
+    <String>[frontendServer, '--sdk-root', sdkRoot, mainPath]
+  );
+
+  String boundaryKey;
+  String outputFilename;
+  server.stderr
+    .transform(UTF8.decoder)
+    .listen((String s) { print(s); });
+  server.stdout
+    .transform(UTF8.decoder)
+    .transform(new LineSplitter())
+    .listen((String string) async {
+      const String RESULT_PREFIX = 'result ';
+      if (boundaryKey == null) {
+        if (string.startsWith(RESULT_PREFIX))
+          boundaryKey = string.substring(RESULT_PREFIX.length);
+      } else {
+        if (string.startsWith(boundaryKey)) {
+          if (string.length > boundaryKey.length)
+            outputFilename = string.substring(boundaryKey.length + 1);
+        }
+        else {
+          print('compile debug message: $string');
+        }
+      }
+    });
+  await server.exitCode;
+  return outputFilename;
+}
+
+class ResidentCompiler {
+  ResidentCompiler(this._sdkRoot) {
+    assert(_sdkRoot != null);
+    if (!_sdkRoot.endsWith('/'))
+      _sdkRoot = "$_sdkRoot/";
+  }
+
+  String _sdkRoot;
+  Process _server;
+  Completer<String> _outputFilename;
+  String _boundaryKey;
+
+  Future<String> recompile(List<String> invalidatedFiles) async {
+    if (_server == null) {
+      throw new Future<String>.error("Recompile should be preceded by compile");
+    }
+
+    _outputFilename = new Completer<String>();
+
+    final String inputKey = new Uuid().generateV4();
+    _server.stdin.writeln('recompile $inputKey');
+    for (String invalidatedFile in invalidatedFiles) {
+      _server.stdin.writeln(invalidatedFile);
+    }
+    _server.stdin.writeln(inputKey);
+
+    return _outputFilename.future;
+  }
+
+  Future<Null> handler(String string) async {
+    const String RESULT_PREFIX = 'result ';
+      if (_boundaryKey == null) {
+        if (string.startsWith(RESULT_PREFIX)) {
+          _boundaryKey = string.substring(RESULT_PREFIX.length);
+        }
+      } else {
+        if (string.startsWith(_boundaryKey)) {
+          _outputFilename.complete(string.length > _boundaryKey.length
+            ? string.substring(_boundaryKey.length + 1)
+            : null
+          );
+          _boundaryKey = null;
+        } else {
+          print('compile debug message: $string');
```

Comment:
here and above, using `print` in flutter_tools is typically a sign of a bug, did you mean printError or printTrace or some such?